Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+                << (((sc.sampaTime + sc.nofSamples() - 1) >= 98) ? " *" : "") << std::endl;
+    }
+
+    // skip channels not associated to any pad
+    if (padId < 0) {
+      return;
+    }
+
+    Digit::Time time;
+    time.sampaTime = sc.sampaTime;
+    time.bunchCrossing = sc.bunchCrossing;
+    time.orbit = orbit;
+
+    Digit::Time stopTime;
+    stopTime.sampaTime = sc.sampaTime + sc.nofSamples() - 1;
+    stopTime.bunchCrossing = sc.bunchCrossing;
+    stopTime.orbit = orbit;
+
+    mMerger->addDigit(feeId, static_cast<int>(dsElecId.solarId()), static_cast<int>(dsElecId.elinkId()), static_cast<int>(channel),
+                      deId, padId, digitadc, time, stopTime);
+    ++ndigits;
+  };
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
+    auto rdhAnyPtr = reinterpret_cast<o2::header::RDHAny*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    if (mRdhHandler) {
+      mRdhHandler(rdhAnyPtr);
+    }
+    auto rdhPtr = reinterpret_cast<o2::header::RAWDataHeaderV4*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    auto& rdh = *rdhPtr;
+    mNrdhs++;
+
+    auto cruId = o2::raw::RDHUtils::getCRUID(*rdhAnyPtr);
+    auto endPointID = o2::raw::RDHUtils::getEndPointID(*rdhAnyPtr);
+    o2::raw::RDHUtils::setFEEID(*rdhAnyPtr, cruId * 2 + endPointID);
+    isStopRDH = o2::raw::RDHUtils::getStop(*rdhAnyPtr);
+
+    feeId = o2::raw::RDHUtils::getFEEID(*rdhAnyPtr);
+    orbit = o2::raw::RDHUtils::getHeartBeatOrbit(*rdhAnyPtr);
+    linkId = o2::raw::RDHUtils::getLinkID(*rdhAnyPtr);
+
+    if (!mMerger) {
+      if (linkId == 15) {
+        mMerger = new Merger;
+      } else {
+        mMerger = new MergerSimple;
+      }
+
+      mMerger->setDigitHandler(storeDigit);
+    }
+    mMerger->setOrbit(feeId, orbit, isStopRDH);
+  };
+
+  if (!mDecoder) {
+    mDecoder = mFee2Solar ? o2::mch::raw::createPageDecoder(page, channelHandler, mFee2Solar)
+                          : o2::mch::raw::createPageDecoder(page, channelHandler);
+  }
+
+  patchPage(page);
+
+  mDecoder(page);
+
+  mMerger->mergeDigits(feeId);
+
+  if (mPrint) {
+    std::cout << "[decodeBuffer] mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    for (auto d : mOutputDigits) {
+      if (d.getPadID() < 0)
+        continue;
+      const Segmentation& segment = segmentation(d.getDetID());
+      float X = segment.padPositionX(d.getPadID());
+      float Y = segment.padPositionY(d.getPadID());
+      bool bend = !segment.isBendingPad(d.getPadID());
+      if (bend)
+        continue;
+      std::cout << fmt::format("  DE {:4d}  PAD {:5d}  ADC {:6d}  TIME ({} {} {:4d})",
+                               d.getDetID(), d.getPadID(), d.getADC(), d.getTime().orbit, d.getTime().bunchCrossing, d.getTime().sampaTime);
+      std::cout << fmt::format("\tC {}  PAD_XY {:+2.2f} , {:+2.2f}", (int)bend, X, Y);
+      std::cout << std::endl;
+    }
+    for (auto d : mOutputDigits) {
```

Comment:
certainly the printing of the bending/non-bending digits can be done without that much duplicated code ? (one more free function, e.g. dumpDigit(gsl::span<Digit> digits, bool bending) to be done ?)

Reply:
The double loop is there to allow printing all the bending digits first, then all the non-bending ones, to simplify the debugging.  I have improved the code with an helper function.